### PR TITLE
Implement mocks in bun:test

### PR DESCRIPTION
**Pull request by Jarred-Sumner**
This implements:
- `expect(fn).toHaveBeenCalled()`
- `expect(fn).toHaveBeenCalledTimes(1)`

```ts
import { describe, expect, it, mock } from "bun:test";

describe("mocks", () => {
  it("are callable", () => {
    const fn = mock(() => 42);
    expect(fn()).toBe(42);
    expect(fn.mock.calls.length).toBe(1);
    expect(fn.mock.calls[0].length).toBe(0);
    expect(fn).toHaveBeenCalled();
    expect(fn).toHaveBeenCalledTimes(1);
  });
});

```

This leaves `toHaveBeenReturned` & friends to a subsequent PR. 

mockImplementationOnce() doesn't work right yet

### CODE REVIEW

File: src/bun.js/test/jest.zig [hunk 1]
```diff
@@ -3821,7 +3829,92 @@ pub const Expect = struct {
...
+        if (not) {
+            const signature = comptime getSignature("toHaveBeenCalled", "<green>expected<r>", true);
+            const fmt = signature ++ "\n\nExpected: not <green>{any}<r>\n";
+            if (Output.enable_ansi_colors) {
+                globalObject.throw(Output.prettyFmt(fmt, true), .{calls.toFmt(globalObject, &formatter)});
+                return .zero;
+            }
+            globalObject.throw(Output.prettyFmt(fmt, false), .{calls.toFmt(globalObject, &formatter)});
+            return .zero;
+        }
```
Comment (dylan-conway) [inline src/bun.js/test/jest.zig:3887]: the not not case needs to be handled here and in `toHaveBeenCalledTimes`

File: src/bun.js/test/jest.zig [hunk 2]
```diff
@@ -3821,7 +3829,92 @@ pub const Expect = struct {
...
+
+        if (arguments.len < 1 or !arguments[0].isAnyInt()) {
+            globalObject.throwInvalidArguments("toHaveBeenCalledTimes() requires 1 integer argument", .{});
+            return .zero;
+        }
+
+        const times = arguments[0].coerce(i32, globalObject);
+
+        var pass = @intCast(i32, calls.getLength(globalObject)) >= times;
```
Comment (dylan-conway) [inline src/bun.js/test/jest.zig]: should this be `==`?

Comment (Jarred-Sumner) [reply src/bun.js/test/jest.zig]: a function that has been called 5 times has also been called 4 times

i didn't test if Jest behaves this way though

Comment (dylan-conway) [reply src/bun.js/test/jest.zig]: just checked, looks like they test for the exact number of times

File: src/bun.js/test/jest.zig [hunk 3]
```diff
@@ -3821,7 +3829,101 @@ pub const Expect = struct {
...
+        const value: JSValue = JSC.Jest.Expect.capturedValueGetCached(thisValue) orelse {
+            globalObject.throw("Internal consistency error: the expect(value) was garbage collected but it should not have been!", .{});
+            return .zero;
+        };
+
+        const calls = JSMockFunction__getCalls(value);
+        active_test_expectation_counter.actual += 1;
+
+        if (calls == .zero or !calls.jsType().isArray()) {
+            globalObject.throw("Expected value must be a mock function: {}", .{value});
```
Comment (Electroid) [inline src/bun.js/test/jest.zig:3857]: The `if` check and error message don't seem to match up?

File: src/bun.js/test/jest.zig [hunk 4]
```diff
@@ -3821,7 +3829,101 @@ pub const Expect = struct {
...
+            const fmt = signature ++ "\n\nExpected <green>{any}<r>\n";
+            if (Output.enable_ansi_colors) {
+                globalObject.throw(Output.prettyFmt(fmt, true), .{calls.toFmt(globalObject, &formatter)});
+                return .zero;
+            }
+            globalObject.throw(Output.prettyFmt(fmt, false), .{calls.toFmt(globalObject, &formatter)});
+            return .zero;
+        }
+
+        unreachable;
```
Comment (Electroid) [inline src/bun.js/test/jest.zig:3945]: Can be removed

File: test/js/bun/test/mock-test.test.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import { test, mock, expect } from "bun:test";
```
Comment (Electroid) [inline test/js/bun/test/mock-test.test.ts]: looks good, would add test cases for:
* async fn that resolves
* async fn that rejects
* fn that returns already resolved promise
* fn called with multiple args
* fn that calls another mock fn

File: src/bun.js/test/jest.zig (hunk 3 above)
Comment (Jarred-Sumner) [reply src/bun.js/test/jest.zig:3857]: that is how it detects if it's a mocked callback

**Pull request merged**